### PR TITLE
Implement tests for ConfirmEmailView

### PR DESCRIPTION
**Pull request by Themanwhosmellslikesugar**

### CODE REVIEW

File: users/tests.py [hunk 1]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_confirm_email_with_invalid_token(self):
+        """Tests if it's not possible to confirm email with invalid token"""
+
+        url = reverse('confirm-email', kwargs={'version': 'v1'})
+        response = self.client.post(url, data=json.dumps({'token': 'random_token'}),
+                                    content_type='application/json')
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (karpa4o4) [inline users/tests.py:485]: Assert response data also.

File: users/tests.py [hunk 2]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+    """Tests the confirm_email endpoint. """
+
+    def setUp(self):
+        super().setUp()
+
+        user = User.objects.get(username=self._user['username'])
+        user.person.email_confirmed = False
+        user.person.save(update_fields=['email_confirmed'])
+
+        self._email_confirmation_token = user.person.emailconfirmationtoken
```
Comment (karpa4o4) [inline users/tests.py]: Get token from `user.person.emailconfirmationtoken.token`.

File: users/tests.py [hunk 3]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+
+    def test_confirm_email(self):
+        """Test if it possible to confirm email. """
+
+        url = reverse('confirm-email', kwargs={'version': 'v1'})
+        data = {'token': str(self._email_confirmation_token)}
```
Comment (karpa4o4) [inline users/tests.py]: Don't cast value to `str`.

File: users/tests.py [hunk 4]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+        super().setUp()
+
+        user = User.objects.get(username=self._user['username'])
+        user.person.email_confirmed = False
+        user.person.save(update_fields=['email_confirmed'])
+
+        self._email_confirmation_token = user.person.emailconfirmationtoken
+
+    def test_confirm_email(self):
+        """Test if it possible to confirm email. """
```
Comment (karpa4o4) [inline users/tests.py]: `Test if it's possible to confirm the user email.`

File: users/tests.py [hunk 5]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+
+        url = reverse('confirm-email', kwargs={'version': 'v1'})
+        data = {'token': str(self._email_confirmation_token)}
+        response = self.client.post(url, data=json.dumps(data),
+                                    content_type='application/json')
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_confirm_email_with_invalid_token(self):
+        """Tests if it's not possible to confirm email with invalid token"""
```
Comment (karpa4o4) [inline users/tests.py]: * `... to confirm the user email ...`;
* `... invalid token. """`

File: users/tests.py [hunk 6]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+
+        url = reverse('confirm-email', kwargs={'version': 'v1'})
+        response = self.client.post(url, data=json.dumps({'token': 'random_token'}),
```
Comment (karpa4o4) [inline users/tests.py]: Let's replace `random_token` to `invalid_token`.

File: users/tests.py [hunk 7]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+        email_confirmation_token.person.save(update_fields=['email_confirmed'])
+        email_confirmation_token.delete()
+
+        url = reverse('confirm-email', kwargs={'version': 'v1'})
+        response = self.client.post(url, data=json.dumps({'token': token}),
+                                    content_type='application/json')
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+        user2.delete()
```
Comment (karpa4o4) [inline users/tests.py]: You don't need to delete the user explicitly.

File: users/tests.py [hunk 8]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+        email_confirmation_token.person.save(update_fields=['email_confirmed'])
+        email_confirmation_token.save(update_fields=['created_at'])
+
+        url = reverse('confirm-email', kwargs={'version': 'v1'})
+        response = self.client.post(url, data=json.dumps({'token': str(token)}),
+                                    content_type='application/json')
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+        user2.delete()
```
Comment (karpa4o4) [inline users/tests.py]: You don't need to delete the user explicitly.

File: users/tests.py [hunk 9]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+        """Tests if it's not possible to confirm email with invalid token"""
+
+        url = reverse('confirm-email', kwargs={'version': 'v1'})
+        response = self.client.post(url, data=json.dumps({'token': 'random_token'}),
+                                    content_type='application/json')
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_confirm_email_with_already_confirmed_email(self):
+        """Tests if it's not possible to confirm email with already confirmed email. """
```
Comment (karpa4o4) [inline users/tests.py]: `... to confirm the user email ...`.

File: users/tests.py [hunk 10]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+        url = reverse('confirm-email', kwargs={'version': 'v1'})
+        response = self.client.post(url, data=json.dumps({'token': token}),
+                                    content_type='application/json')
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+        user2.delete()
+
+    def test_confirm_email_with_expired_token(self):
+        """Tests if it's not possible to confirm email with expired token. """
```
Comment (karpa4o4) [inline users/tests.py]: `... to confirm the user email ...`.

File: users/tests.py [hunk 11]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+        email_confirmation_token = EmailConfirmationToken.objects.get(token=token)
+        email_confirmation_token.person.email_confirmed = True
+        email_confirmation_token.person.save(update_fields=['email_confirmed'])
+        email_confirmation_token.delete()
+
+        url = reverse('confirm-email', kwargs={'version': 'v1'})
+        response = self.client.post(url, data=json.dumps({'token': token}),
+                                    content_type='application/json')
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (karpa4o4) [inline users/tests.py:499]: Assert response data also.

File: users/tests.py [hunk 12]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+        email_confirmation_token.created_at = timezone.now() - timedelta(
+            minutes=settings.EMAIL_CONFIRMATION_TOKEN_TTL)
+        email_confirmation_token.person.save(update_fields=['email_confirmed'])
+        email_confirmation_token.save(update_fields=['created_at'])
+
+        url = reverse('confirm-email', kwargs={'version': 'v1'})
+        response = self.client.post(url, data=json.dumps({'token': str(token)}),
+                                    content_type='application/json')
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (karpa4o4) [inline users/tests.py:518]: Assert response data also.

File: users/tests.py [hunk 13]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+        """Tests if it's not possible to confirm email with already confirmed email. """
+
+        user2 = {
+            'username': 'test.user2',
+            'password': 'secret',
+            'email': 'test.user2@domain.com',
+        }
+        user2 = User.objects.create_user(**user2)
+
+        token = str(user2.person.emailconfirmationtoken)
```
Comment (karpa4o4) [inline users/tests.py:500]: Don't create an additional user, use user created in `BaseSingleUserTest` class.

File: users/tests.py [hunk 14]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+        """Tests if it's not possible to confirm email with expired token. """
+
+        user2 = {
+            'username': 'test.user2',
+            'password': 'secret',
+            'email': 'test.user2@domain.com',
+        }
+        user2 = User.objects.create_user(**user2)
+
+        token = str(user2.person.emailconfirmationtoken)
```
Comment (karpa4o4) [inline users/tests.py:524]: Don't create an additional user, use user created in `BaseSingleUserTest` class.

File: users/tests.py [hunk 15]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+
+    def test_confirm_email_with_already_confirmed_email(self):
+        """Tests if it's not possible to confirm email with already confirmed email. """
+
+        user2 = {
+            'username': 'test.user2',
+            'password': 'secret',
+            'email': 'test.user2@domain.com',
+        }
+        user2 = User.objects.create_user(**user2)
```
Comment (karpa4o4) [inline users/tests.py]: Don't create an additional user, use user created in `BaseSingleUserTest` class.

File: users/tests.py [hunk 16]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+
+    def test_confirm_email_with_expired_token(self):
+        """Tests if it's not possible to confirm email with expired token. """
+
+        user2 = {
+            'username': 'test.user2',
+            'password': 'secret',
+            'email': 'test.user2@domain.com',
+        }
+        user2 = User.objects.create_user(**user2)
```
Comment (karpa4o4) [inline users/tests.py]: Don't create an additional user, use user created in `BaseSingleUserTest` class.

File: users/tests.py [hunk 17]
```diff
@@ -450,3 +454,85 @@ def test_user_profile_delete_unauthorized(self):
...
+
+
+class UserConfirmEmailTest(BaseSingleUserTest):
+    """Tests the confirm_email endpoint. """
+
+    def setUp(self):
+        super().setUp()
+
+        user = User.objects.get(username=self._user['username'])
+        user.person.email_confirmed = False
```
Comment (karpa4o4) [inline users/tests.py:465]: Set `email_confirmed` to `False` only in `test_confirm_email` and `test_confirm_email_with_invalid_token` test methods.

File: users/tests.py [hunk 18]
```diff
@@ -450,3 +453,69 @@ def test_user_profile_delete_unauthorized(self):
...
+
+        self.assertEqual(response.content,
+                         b'{"token": ["Ensure this field has no more than 8 characters."]}')
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_confirm_email_with_already_confirmed_email(self):
+        """Tests if it's not possible to confirm the user email with already confirmed email. """
+
+        user = User.objects.get(username=self._user['username'])
+        user.person.email_confirmed = True
```
Comment (karpa4o4) [inline users/tests.py]: You don't need to set the `email_confirmed` to `True`, it already has this value.

File: users/tests.py [hunk 19]
```diff
@@ -450,3 +453,69 @@ def test_user_profile_delete_unauthorized(self):
...
+        """Tests if it's not possible to confirm the user email with expired token. """
+
+        user = User.objects.get(username=self._user['username'])
+        user.person.email_confirmed = False
+
+        email_confirmation_token = user.person.emailconfirmationtoken.token
+        user.person.emailconfirmationtoken.created_at = timezone.now() - timedelta(
+            minutes=settings.EMAIL_CONFIRMATION_TOKEN_TTL)
+
+        user.person.save(update_fields=['email_confirmed'])
```
Comment (karpa4o4) [inline users/tests.py]: Move instructions after `user.person.email_confirmed = False`.

**Review by karpa4o4 — DISMISSED** [review]

File: users/tests.py [hunk 20]
```diff
@@ -450,3 +453,66 @@ def test_user_profile_delete_unauthorized(self):
...
+        url = reverse('confirm-email', kwargs={'version': 'v1'})
+        response = self.client.post(url, data=json.dumps({'token': 'invalid_token'}),
+                                    content_type='application/json')
+
+        self.assertEqual(response.content,
+                         b'{"token": ["Ensure this field has no more than 8 characters."]}')
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_confirm_email_with_already_confirmed_email(self):
+        """Tests if it's not possible to confirm the user email with already confirmed email. """
```
Comment (eugulixes) [inline users/tests.py]: `Tests if it's not possible to confirm an already confirmed email.`

File: users/tests.py [hunk 21]
```diff
@@ -450,3 +453,66 @@ def test_user_profile_delete_unauthorized(self):
...
+
+        url = reverse('confirm-email', kwargs={'version': 'v1'})
+        response = self.client.post(url, data=json.dumps({'token': email_confirmation_token}),
+                                    content_type='application/json')
+
+        self.assertEqual(response.content, b'{"token": ["Token doesn\'t exist."]}')
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_confirm_email_with_expired_token(self):
+        """Tests if it's not possible to confirm the user email with expired token. """
```
Comment (eugulixes) [inline users/tests.py]: `...with an expired token.`

**Review by karpa4o4 — APPROVED** [review]

**Pull request merged**